lix/packages/sdk/src/resolve-conflict/resolve-conflict.test.ts

```diff
@@ -0,0 +1,460 @@
+		.execute();
+
+	const changes = await lix.db
+		.insertInto("change")
+		.values(mockChanges)
+		.returningAll()
+		.execute();
+
+	const conflict = await lix.db
+		.insertInto("conflict")
+		.values({
+			change_id: changes[0]!.id,
+			conflicting_change_id: changes[1]!.id,
+		})
+		.returningAll()
+		.executeTakeFirstOrThrow();
+
+	await resolveConflict({
+		lix: lix,
+		conflict: conflict,
+		resolveWithChange: changes[0]!,
+	});
+
+	const resolvedConflict = await lix.db
+		.selectFrom("conflict")
+		.selectAll()
+		.where("change_id", "=", conflict.change_id)
+		.where("conflicting_change_id", "=", conflict.conflicting_change_id)
+		.executeTakeFirstOrThrow();
+
+	const fileAfterResolve = await lix.db
+		.selectFrom("file")
+		.selectAll()
+		.where("id", "=", changes[0]!.file_id)
+		.executeTakeFirstOrThrow();
+
+	const parsed = new TextDecoder().decode(fileAfterResolve.data);
+
+	expect(parsed).toBe(mockChanges[0]!.value);
+	expect(resolvedConflict.resolved_with_change_id).toBe(changes[0]!.id);
+});
+
+test("it should throw if the to be resolved with change already exists in the database but is not equal (change immutability is violated)", async () => {
+	const mockChanges: Change[] = [
+		{
+			id: "change1",
+			operation: "create",
+			plugin_key: "plugin1",
+			type: "mock",
+			file_id: "mock",
+			// @ts-expect-error - manual stringification
+			value: JSON.stringify({
+				key: "value1",
+			}),
+		},
+		{
+			id: "change2",
+			operation: "create",
+			file_id: "mock",
+			plugin_key: "plugin1",
+			type: "mock",
+			// @ts-expect-error - manual stringification
+			value: JSON.stringify({
+				key: "value2",
+			}),
+		},
+	];
+
+	const mockPlugin: LixPlugin = {
+		key: "plugin1",
+		glob: "*",
+		applyChanges: vi.fn().mockResolvedValue({
+			fileData: new TextEncoder().encode(
+				mockChanges[0]?.value as unknown as string,
+			),
+		}),
+		diff: {
+			file: vi.fn(),
+		},
+	};
+
+	const lix = await openLixInMemory({
+		blob: await newLixFile(),
+		providePlugins: [mockPlugin],
+	});
+
+	await lix.db
+		.insertInto("file")
+		.values({ id: "mock", path: "mock", data: new Uint8Array() })
+		.execute();
+
+	const changes = await lix.db
+		.insertInto("change")
+		.values(mockChanges)
+		.returningAll()
+		.execute();
+
+	const conflict = await lix.db
+		.insertInto("conflict")
+		.values({
+			change_id: changes[0]!.id,
+			conflicting_change_id: changes[1]!.id,
+		})
+		.returningAll()
+		.executeTakeFirstOrThrow();
+
+	await expect(
+		resolveConflict({
+			lix: lix,
+			conflict: conflict,
+			resolveWithChange: {
+				...changes[0]!,
+				// @ts-expect-error - manual stringification
+				value: JSON.stringify({
+					key: "mutated-value",
+				}),
+			},
+		}),
+	).rejects.toThrowError(new ChangeHasBeenMutatedError());
+});
+
+// the sequence of changes will be broken otherwise
+test("resolving a conflict should throw if the to be resolved with change is not a direct child of the conflicting changes", async () => {
```

Comment:
i assume you mean not "direct child", but  just any child? why would a conflict not be resolvable by an edit that is made to child of a conflict?

Reply:
Good one. It seems to depend on the sequence implementation. The direct child test assumes implementing a sequence as described in version 2. Let's hold of with the discussion until we touch https://github.com/opral/inlang-sdk/issues/143

**1. A sequence is never re-written (changes only appended)**

_Before resolving_

```
changes = [ 1, 2, 3, 4 ] 
sequence = [ 1, 2, 3 ]
conflicts = [ 2  with 4 ]
```

_After resolving by choosing change 4 as the winner_


```
changes = [ 1, 2, 3, 4 ] 
sequence = [ 1, 2, 3, 4 ]
conflicts = [ ]
```

**2. A sequence can be re-written "pick and choose the changes you are about"**

_Before resolving_

```
changes = [ 1, 2, 3, 4 ] 
sequence = [ 1, 2, 3 ]
conflicts = [ 2  with 4 ]
```

_After resolving (4 is the winner, sequence swaps 2 for 4)_

```
changes = [ 1, 2, 3, 4 ] 
sequence = [ 1, 4, 3 ]
conflicts = [ ]
```